runtime/parachains/src/disputes/slashing.rs

```diff
@@ -0,0 +1,905 @@
+	let is_known_offence = match dispute_proof.kind {
+		SlashingOffenceKind::ForInvalid =>
+			<T::HandleReports as HandleReports<T>>::is_known_for_invalid_offence(
+				&[offender],
+				&dispute_proof.time_slot,
+			),
+		SlashingOffenceKind::AgainstValid =>
+			<T::HandleReports as HandleReports<T>>::is_known_against_valid_offence(
+				&[offender],
+				&dispute_proof.time_slot,
+			),
+	};
+
+	if is_known_offence {
+		Err(InvalidTransaction::Stale.into())
+	} else {
+		Ok(())
+	}
+}
+
+/// Actual `HandleReports` implemention.
+///
+/// When configured properly, should be instantiated with
+/// `T::KeyOwnerIdentification, Offences, ReportLongevity`
+/// parameters.
+pub struct SlashingReportHandler<I, R, L> {
+	_phantom: sp_std::marker::PhantomData<(I, R, L)>,
+}
+
+impl<I, R, L> Default for SlashingReportHandler<I, R, L> {
+	fn default() -> Self {
+		Self { _phantom: Default::default() }
+	}
+}
+
+impl<T, R, L> HandleReports<T> for SlashingReportHandler<T::KeyOwnerIdentification, R, L>
+where
+	T: Config + frame_system::offchain::SendTransactionTypes<Call<T>>,
+	R: ReportOffence<
+			AccountId<T>,
+			T::KeyOwnerIdentification,
+			ForInvalidOffence<T::KeyOwnerIdentification>,
+		> + ReportOffence<
+			AccountId<T>,
+			T::KeyOwnerIdentification,
+			AgainstValidOffence<T::KeyOwnerIdentification>,
+		>,
+	L: Get<u64>,
+{
+	type ReportLongevity = L;
+
+	fn report_for_invalid_offence(
+		reporters: Vec<AccountId<T>>,
+		offence: ForInvalidOffence<T::KeyOwnerIdentification>,
+	) -> Result<(), OffenceError> {
+		R::report_offence(reporters, offence)
+	}
+
+	fn report_against_valid_offence(
+		reporters: Vec<AccountId<T>>,
+		offence: AgainstValidOffence<T::KeyOwnerIdentification>,
+	) -> Result<(), OffenceError> {
+		R::report_offence(reporters, offence)
+	}
+
+	fn is_known_for_invalid_offence(
+		offenders: &[T::KeyOwnerIdentification],
+		time_slot: &DisputesTimeSlot,
+	) -> bool {
+		<R as ReportOffence<
+			AccountId<T>,
+			T::KeyOwnerIdentification,
+			ForInvalidOffence<T::KeyOwnerIdentification>,
+		>>::is_known_offence(offenders, time_slot)
+	}
+
+	fn is_known_against_valid_offence(
+		offenders: &[T::KeyOwnerIdentification],
+		time_slot: &DisputesTimeSlot,
+	) -> bool {
+		<R as ReportOffence<
+			AccountId<T>,
+			T::KeyOwnerIdentification,
+			AgainstValidOffence<T::KeyOwnerIdentification>,
+		>>::is_known_offence(offenders, time_slot)
+	}
+
+	fn submit_unsigned_slashing_report(
+		dispute_proof: DisputeProof,
+		key_owner_proof: <T as Config>::KeyOwnerProof,
+	) -> DispatchResult {
+		use frame_system::offchain::SubmitTransaction;
+
+		let session_index = dispute_proof.time_slot.session_index;
+		let validator_index = dispute_proof.validator_index.0;
+		let kind = dispute_proof.kind;
+
+		let call = Call::report_dispute_lost_unsigned {
+			dispute_proof: Box::new(dispute_proof),
+			key_owner_proof,
+		};
+
+		match SubmitTransaction::<T, Call<T>>::submit_unsigned_transaction(call.into()) {
+			Ok(()) => log::info!(
```

Comment:
This should be debug level right ?

Reply:
This will be only printed on a block author node that will produce these reports. I think it's helpful to be info and it's not expected to be spammy at all.